tools/crm_ticket.c

```diff
@@ -427,50 +427,107 @@ find_ticket_constraints(cib_t * the_cib, gchar *ticket_id, xmlNode ** ticket_con
 
-    if (state_xml == NULL) {
-        return rc;
-    }
+    pcmk__formatted_printf(out, "Constraints XML:\n\n");
+    out->output_xml(out, PCMK_XE_CONSTRAINT, s);
+    free(s);
+    return pcmk_rc_ok;
+}
 
-    fprintf(stdout, "State XML:\n");
-    if (state_xml) {
-        char *state_xml_str = NULL;
+PCMK__OUTPUT_ARGS("ticket-constraints", "gchar *", "xmlNode *")
+static int
+ticket_constraints_xml(pcmk__output_t *out, va_list args)
+{
+    gchar *ticket_id = va_arg(args, gchar *);
+    xmlNode *constraint_xml = va_arg(args, xmlNode *);
+    const char *rsc = NULL;
+
+    char *s = dump_xml_formatted(constraint_xml);
+
+    /* Create:
+     * <tickets>
+     *   <ticket id="">
+     *     <constraints>
+     *       <rsc_ticket />
+     *     </constraints>
+     *   </ticket>
+     * </tickets>
+     */
+    pcmk__output_xml_create_parent(out, PCMK_XE_TICKETS, NULL);
+    pcmk__output_xml_create_parent(out, PCMK_XE_TICKET,
+                                   PCMK_XA_ID, ticket_id, NULL);
+    pcmk__output_xml_create_parent(out, PCMK_XE_CONSTRAINTS, NULL);
+    pcmk__output_xml_add_node_copy(out, constraint_xml);
+
+    /* Pop three parents so we are now back at the same level where
+     * the <tickets> element was created.
+     */
+    pcmk__output_xml_pop_parent(out);
+    pcmk__output_xml_pop_parent(out);
+    pcmk__output_xml_pop_parent(out);
+
+    /* Constraints can reference a resource ID that is defined in the XML
+     * schema as an IDREF.  This requires some other element to be present
+     * with an id= attribute that matches.
+     *
+     * Create the following:
+     * <resources>
+     *   <resource id="" />
+     * </resources>
+     */
+    rsc = crm_element_value(constraint_xml, PCMK_XA_RSC);
 
-        state_xml_str = dump_xml_formatted(state_xml);
-        fprintf(stdout, "\n%s", state_xml_str);
-        free_xml(state_xml);
-        free(state_xml_str);
-    }
+    pcmk__output_xml_create_parent(out, PCMK_XE_RESOURCES, NULL);
+    pcmk__output_xml_create_parent(out, PCMK_XE_RESOURCE,
+                                   PCMK_XA_ID, rsc, NULL);
 
-    return rc;
+    free(s);
+    return pcmk_rc_ok;
 }
 
+PCMK__OUTPUT_ARGS("ticket-state", "gchar *", "xmlNode *")
 static int
-dump_constraints(cib_t * the_cib, gchar *ticket_id)
+ticket_state_default(pcmk__output_t *out, va_list args)
 {
-    int rc = pcmk_rc_ok;
-    xmlNode *cons_xml = NULL;
-    char *cons_xml_str = NULL;
+    gchar *ticket_id G_GNUC_UNUSED = va_arg(args, gchar *);
+    xmlNode *state_xml = va_arg(args, xmlNode *);
 
-    rc = find_ticket_constraints(the_cib, ticket_id, &cons_xml);
+    char *s = dump_xml_formatted(state_xml);
 
-    if (cons_xml == NULL) {
-        return rc;
-    }
-
-    cons_xml_str = dump_xml_formatted(cons_xml);
-    fprintf(stdout, "Constraints XML:\n\n%s", cons_xml_str);
-    free_xml(cons_xml);
-    free(cons_xml_str);
+    pcmk__formatted_printf(out, "State XML:\n\n");
+    out->output_xml(out, PCMK__XE_TICKET_STATE, s);
```

Comment:
If backward compatibility weren't an issue, it would be nice to implement friendly text output